File: plugins/process-monitor/process-monitor.10s.py
#!/usr/bin/env python3
"""Process Monitor - Shows top CPU/Memory processes"""
import subprocess
import platform

def get_top_processes():
    try:
        if platform.system() == 'Darwin':
            # macOS
            cmd = "ps aux | sort -nrk 3 | head -6 | tail -5"
        else:
            # Linux
            cmd = "ps aux --sort=-%cpu | head -6 | tail -5"

        result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
        lines = result.stdout.strip().split('\n')

        processes = []
        for line in lines:
            parts = line.split()
            if len(parts) >= 11:
                cpu = parts[2]
                mem = parts[3]
                cmd = ' '.join(parts[10:])[:30]
                processes.append({'cpu': cpu, 'mem': mem, 'cmd': cmd})

        return processes
    except:
        return []

processes = get_top_processes()
total_processes = subprocess.run(['sh', '-c', 'ps aux | wc -l'], capture_output=True, text=True)
count = int(total_processes.stdout.strip()) - 1

print(f" {count}")
print("---")
print(f"Running Processes: {count}")
print("---")
print("Top by CPU:")

for p in processes:
    cpu = float(p['cpu'])
    color = 'red' if cpu > 50 else 'yellow' if cpu > 20 else 'green'
    print(f"  {p['cpu']}% {p['cmd']} | color={color}")

print("---")
print("Open Activity Monitor | bash='open -a \"Activity Monitor\"' terminal=false")
print("Refresh | refresh=true")

File: plugins/process-monitor/process-monitor.10s.lua
-- process-monitor.10s.lua
-- Process Monitor (desktop only)

local function env(name, default)
    local value = crossbar.env(name, default)
    if value == nil or value == '' then
        return default
    end
    return value
end

local function env_num(name, default)
    local value = env(name, tostring(default))
    local num = tonumber(value)
    if num == nil then
        return default
    end
    return num
end

local is_mobile = crossbar.isMobile()
local max_items = env_num('PROC_MAX_ITEMS', 5)
if max_items < 1 then max_items = 1 end
if max_items > 50 then max_items = 50 end
if is_mobile then
    print('PROC N/A | color=gray')
    print('---')
    print('Process monitoring is desktop-only')
    print('---')
    print('Refresh | refresh=true')
    return
end

local platform = crossbar.platform()
local cmd = ''
if platform == 'macos' then
    cmd = 'ps aux | sort -nrk 3 | head -6'
else
    cmd = 'ps aux --sort=-%cpu | head -6'
end

local output = crossbar.exec(cmd)
local total_proc_raw = crossbar.exec('ps aux | wc -l')
local total_proc = tonumber((total_proc_raw or ''):match('%d+')) or 0
if total_proc > 0 then total_proc = total_proc - 1 end

print('PROC ' .. tostring(total_proc))
print('---')
print('Running Processes: ' .. tostring(total_proc))
print('---')
print('Top by CPU:')

local shown = 0
for line in (output or ''):gmatch('[^\n]+') do
    if not line:match('^USER') then
        local parts = {}
        for part in line:gmatch('%S+') do
            table.insert(parts, part)
        end

        if #parts >= 11 then
            local cpu = tonumber(parts[3]) or 0
            local proc_cmd = ''
            for i = 11, #parts do
                proc_cmd = proc_cmd .. parts[i] .. ' '
            end
            proc_cmd = proc_cmd:sub(1, 30)

            local color = 'green'
            if cpu > 50 then color = 'red'
            elseif cpu > 20 then color = 'yellow'
            end

            print(string.format('  %.1f%% %s | color=%s', cpu, proc_cmd, color))
            shown = shown + 1
            if shown >= max_items then
                break
            end
        end
    end
end

if shown == 0 then
    print('No process data available')
end

print('---')
print('Refresh | refresh=true')
